rasa/utils/tensorflow/layers.py

```diff
@@ -640,6 +648,67 @@ def _get_bad_mask(
             tf.reduce_all(tf.equal(neg_labels, pos_labels), axis=-1), pos_labels.dtype
         )
 
+    def _get_same_label_mask(
+        self, labels_a: tf.Tensor, labels_b: tf.Tensor
+    ) -> tf.Tensor:
+
+        target_labels_expanded = tf.tile(
+            tf.expand_dims(labels_b, 0), (tf.shape(labels_a)[0], 1, 1)
+        )
+        same_labels_mask = tf.not_equal(
+            tf.expand_dims(labels_a, 1), target_labels_expanded
+        )
+        return tf.cast(same_labels_mask, dtype=tf.bool)
+
+    def _get_similarity_to_optimize(
+        self,
+        embeds_a: tf.Tensor,
+        embeds_b: tf.Tensor,
+        labels_a: tf.Tensor,
+        labels_b: tf.Tensor,
+        mask: tf.Tensor,
+    ) -> tf.Tensor:
+        """Sample negatives from `embeddings_b` based on similarity between `embeddings_a` and `embeddings_b`.
+        
+        Args:
+            embeds_a: First set of embeddings in the pair. (bs x seq x dim)
+            embeds_b: Second set of embeddings in the pair. (ns x seq x dim)
+            labels_a: Labels for first set of embeddings in the pair. (bs x seq x 1)
+            labels_b: Labels for second set of embeddings in the pair. (ns x seq x 1)
+
+        Returns:
+            Sampled negatives, mask indicating bad negatives, computed similarity between a and b
+        """
+        embeds_a_flat = self._make_flat(embeds_a)
+        embeds_b_flat = self._make_flat(embeds_b)
+        labels_a_flat = self._make_flat(labels_a)
+        labels_b_flat = self._make_flat(labels_b)
+
+        # Compute the mask to indicate which pairs of
+        # inputs have a different label. The mask will have 0
+        # for same label and 1 for different label.
+        unique_label_mask = self._get_same_label_mask(
+            labels_a_flat, labels_b_flat
+        )  # (bs x ns x 1)
+
+        # Similarity between each unique pair of embeddings.
+        # Pair is composed of elements from first embeddings and
+        # elements from second embeddings.
+        pairwise_similarities = self.sim(
+            tf.expand_dims(embeds_a_flat, 1), tf.expand_dims(embeds_b_flat, 0), mask
```

Comment:
Axis of expand dims need to be checked here for TED as sequence length is also a dimension.